File: spotifyarch/src/main/java/com/timothy/spotifyarchitecture/retrofit/SpotifyCreator.java
package com.timothy.spotifyarchitecture.retrofit;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by tim on 5/28/17.
 */

public class SpotifyCreator extends Spotify {

    public static SpotifyService createAuthenticatedService(ApiAuthenticator authenticator) {
        Retrofit build = (new Retrofit.Builder()).client(createHttpClient(authenticator)).addConverterFactory(GsonConverterFactory.create()).baseUrl("https://api.spotify.com/v1/").build();
        return build.create(SpotifyService.class);
    }

    public static SpotifyService createNotAuthenticatedService() {
        Retrofit build = (new Retrofit.Builder()).client(createNoAuthHttpClient()).addConverterFactory(GsonConverterFactory.create()).baseUrl("https://api.spotify.com/v1/").build();
        return build.create(SpotifyService.class);
    }

    public static SpotifyServiceAuth createServiceAuthenticator() {
        Retrofit build = (new Retrofit.Builder()).client(createNoAuthHttpClient()).addConverterFactory(GsonConverterFactory.create()).baseUrl(SpotifyServiceAuth.SPOTIFY_AUTH_API_ENDPOINT).build();
        return build.create(SpotifyServiceAuth.class);
    }

    public static OkHttpClient createHttpClient(ApiAuthenticator authenticator) {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BASIC);
        return (new okhttp3.OkHttpClient.Builder()).addInterceptor(interceptor).addInterceptor(authenticator).build();
    }

    public static OkHttpClient createNoAuthHttpClient() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BASIC);
        return (new okhttp3.OkHttpClient.Builder()).addInterceptor(interceptor).build();
    }

    public static ApiAuthenticator createApiAuthenticator(String token) {
        return new ApiAuthenticator(token);
    }

    public static class ApiAuthenticator implements Interceptor {
        private String mAccessToken;

        public ApiAuthenticator(String accessToken) {
            this.mAccessToken = accessToken;
        }

        public void setAccessToken(String accessToken) {
            this.mAccessToken = accessToken;
        }

        public Response intercept(Interceptor.Chain chain) throws IOException {
            Request request = chain.request();
            if(this.mAccessToken != null) {
                Request authRequest = request.newBuilder().addHeader("Authorization", "Bearer " + this.mAccessToken).build();
                return chain.proceed(authRequest);
            } else {
                return chain.proceed(request);
            }
        }
    }

}

File: spotify-web/src/main/kotlin/com/timmahh/spotifyweb/retrofit/SpotifyCreator.kt
package com.timmahh.spotifyweb.retrofit

import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException

/**
 * A utility class to create retrofit services to access the Spotify Web API
 */

class SpotifyCreator : Spotify() {

    class ApiAuthenticator(var accessToken: String = "") : Interceptor {

        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): Response {
            val request = chain.request()
            if (!this.accessToken.isEmpty()) {
                val authRequest = request.newBuilder().addHeader("Authorization", "Bearer " + this.accessToken).build()
                return chain.proceed(authRequest)
            } else {
                return chain.proceed(request)
            }
        }
    }

    companion object {

        fun createAuthenticatedService(authenticator: ApiAuthenticator): SpotifyService {
            val build = Retrofit.Builder().client(createHttpClient(authenticator)).addConverterFactory(GsonConverterFactory.create()).baseUrl("https://api.spotify.com/v1/").build()
            return build.create(SpotifyService::class.java)
        }

       /* fun createNotAuthenticatedService(): SpotifyService {
            val build = Retrofit.Builder().client(createNoAuthHttpClient()).addConverterFactory(GsonConverterFactory.create()).baseUrl("https://api.spotify.com/v1/").build()
            return build.create(SpotifyService::class.java)
        }*/

        fun createServiceAuthenticator(): SpotifyServiceAuth {
            val build = Retrofit.Builder().client(createNoAuthHttpClient()).addConverterFactory(GsonConverterFactory.create()).baseUrl(SpotifyServiceAuth.SPOTIFY_AUTH_API_ENDPOINT).build()
            return build.create(SpotifyServiceAuth::class.java)
        }

        fun createHttpClient(authenticator: ApiAuthenticator): OkHttpClient {
            val interceptor = HttpLoggingInterceptor()
            interceptor.level = HttpLoggingInterceptor.Level.BASIC
            return okhttp3.OkHttpClient.Builder().addInterceptor(interceptor).addInterceptor(authenticator).build()
        }

        fun createNoAuthHttpClient(): OkHttpClient {
            val interceptor = HttpLoggingInterceptor()
            interceptor.level = HttpLoggingInterceptor.Level.BASIC
            return okhttp3.OkHttpClient.Builder().addInterceptor(interceptor).build()
        }

        fun createApiAuthenticator(token: String = ""): ApiAuthenticator {
            return ApiAuthenticator(token)
        }
    }

}
